### ISSUE TITLE
check a GADSdat for compliance with requirements for Stata data sets

### ISSUE THREAD

**Issue by weigtnik** [beckerbenj/eatGADS#110]
This is basically a copy of [issue #25 in eatFDZ](https://github.com/beckerbenj/eatFDZ/issues/25#issue-2638810504). Some characteristics of (meta) data can cause trouble when a `GADSdat` is being exported into a `.dta` file. These issues lead to warnings and errors, e.g. from `haven`, that might not be too helpful to the user. Therefore, a new wrapper function, `check4Stata`, should check for the following potential problems.

- [x] (1) dots in variable names (e.g., `IDstud.new`) --> implemented with `checkVarNames(..., checkDots = TRUE)`
- [x] (2) special characters in variable names, variable labels, and value labels --> implemented with `fixEncoding()`, which will probably be expanded in the future (#86)
- [x] (3) labeled character variables --> implemented with `checkGADSdat()` and somewhat difficult to even "achieve" in a `GADSdat` (see comments below)
- [x] (4) labeled fractional values --> new function `checkLabeledFractionals()`
- [x] (5) labeled extreme values exceeding R's maximum capacity for integers at around 2 * 10^9 --> new function `checkIntOverflow()`
- [x] (6) long names, labels, and character values --> new joint function for `SPSS` and `Stata`, as @beckerbenj suggested (tbd together with @sachseka this week?)

The return value should be similar to `check4SPSS()`. Previously suggested:
- Dots (1) and spec. char.s (2) should be fixed automatically (as they already are when using the referenced functions).
- For labeled `char` values (3), I suggest the user choose whether it should be fixed automatically.
- Fixing lab. fract. val.s (4) should remain in the hands of users for now. Suggestions on how to fix them could be added later.
- Extreme values (5) should also remain in the hands of users for now. Automatic recoding/"cutting down" of values that clearly are typos (e.g., a missing-tagged 9999999999 when the next lower missing is at 9998) could be added later and dependent on users' choice.
- For long characters (6), I suggest leaving names and values unchanged and cutting labels.

**Comment by beckerbenj** [comment]
Thanks for setting up the issue and taking this over, looks great! Just some unordered thoughts:

> (3) labeled character variables --> implement in check4Stata()directly, maybe using char_valLabels2numeric, becuase checkGADSdat() will throw an error giving only limited information

I think currently it is very tricky to import data in any way that allows labeled character values. Simply speaking, `eatGADS` itself does not support labeled characters (except numeric missing tags?). We could check whether there are any obvious gaps I may missed; but if this is the case, maybe `check_GADSdat()` can suffice. (see also `labeledStrings` argument in `import_spss()`

> (6) long names, labels, and character values --> implement in check4Stata() directly because it is Stata-specific

While the specific lengths may be Stata specific, maybe common functions for Stata and SPSS could be written with an argument that allows setting the tolerated length?

**Comment by weigtnik** [comment]
On labeled characters: That's a good point. It would only really be possible to end up with labeled characters if `GADSdat$labels` was manipulated directly, right? Then, I agree that the error `check_GADSdat()` would throw is sufficient to remind users of their wrongdoing.

On extensive lengths: I agree that this would be a good idea! Would it make sense to also include a check on file path lengths in that function? I ran into that problem yesterday with our beloved folder structure. 😅

In general: On second thoughts, I would like to change my suggestion on the behavior of the wrapper. Maybe it would make more sense to have it only *check* and report on compliance but not change the `GADSdat` and instead return `TRUE` or `FALSE` depending on whether the checks are passed.

**Comment by beckerbenj** [comment]
A check regarding file path length is probably always a good idea (on Windows...). However, this would be something for the `write` functions themselves, as `check4Stata()` will not be dealing with any paths, or will it?

Regarding its output: It would be great, if `check4Stata()` could follow the output of the existing `check` functions in `eatGADS` (it would be confusing for users if, for instance, `check4spss()` and `check4stata()` work in very different ways). Maybe you could check how `check4spss()` works (which might be already close to your last suggestion?)?

**Comment by weigtnik** [comment]
> However, this would be something for the write functions themselves, as check4Stata() will not be dealing with any paths, or will it?

Absolutely. I was just thinking another step ahead and about adding one function combining both/all three purposes of checking for SPSS, Stata, and file paths. However, that probably was one step too far because a file path check should reasonably return a rather different kind of information than a metadata check.


> It would be great, if `check4Stata()` could follow the output of the existing `check` functions in `eatGADS`

I agree, and just like you said, `check4SPSS()` does seem very close to what I was thinking about. I will aim to write something as similar as possible!

**Comment by weigtnik** [comment]
I would suggest adding a prefix to the functions that would most likely only be used before exporting a data set as `.sav` or `.dta`. This mainly pertains to the new auxiliary functions I am writing, but also to `checkFormat` and potentially `check4SPSS`, too - depending on whether or not we decide that the wrapper functions should also get that prefix as well.
So, my questions for discussion are:
1. Should the auxiliary functions in the "data compliance check framework" get a prefix? Maybe `dcc` for data compliance check (= `dccLabeledFractionals`)?
2. Alternatively, should the auxiliary functions be exported at all since they might almost never be called outside the wrapper functions? In that case, the prefix may be unnecessary.
3. Would it be okay that an existing function (`checkFormat` and/or `checkSPSS`) is either renamed or booted from the index?
4. Should the wrappers get a prefix?

**Comment by beckerbenj** [comment]
Conceptual question:
Should `check4SPSS()` and `check4Stata()` be included in `eatFDZ::check_all()`?
